src/main/java/ch/njol/skript/expressions/ExprName.java

```diff
@@ -97,336 +96,151 @@
-			return new Class[0];
-		if (mode != ChangeMode.SET)
-			return null;
-		if ((type.acceptChange & PLAYER) != 0 && Player.class.isAssignableFrom(getExpr().getReturnType())) {
-			changeType = PLAYER;
-		} else if ((type.acceptChange & INVENTORY) != 0 && Inventory.class.isAssignableFrom(getExpr().getReturnType())) {
-			changeType = INVENTORY;
-		} else if ((type.acceptChange & ITEM) != 0 && (getExpr().isSingle() && ChangerUtils.acceptsChange(getExpr(), ChangeMode.SET, ItemStack.class, ItemType.class) || Slot.class.isAssignableFrom(getExpr().getReturnType()))) {
-			changeType = ITEM;
-		} else if ((type.acceptChange & ENTITY) != 0 && Entity.class.isAssignableFrom(getExpr().getReturnType())) {
-			if (type == NameType.NAME && Player.class.isAssignableFrom(getExpr().getReturnType())) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.RESET) {
+			if (mark == 1 && Player.class.isAssignableFrom(getExpr().getReturnType())) {
 				Skript.error("Can't change the Minecraft name of a player. Change the 'display name' or 'tab list name' instead.");
 				return null;
 			}
-			changeType = ENTITY;
+			return CollectionUtils.array(String.class);
 		}
-		return changeType == 0 ? null : CollectionUtils.array(String.class);
+		return null;	
 	}
 
 	@Override
 	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
-		String name = delta == null ? null : (String) delta[0];
-		if (changeType == ITEM) {
-			if (Slot.class.isAssignableFrom(getExpr().getReturnType())) {
-				for (Slot s : (Slot[]) getExpr().getArray(e)) {
-					ItemStack i = s.getItem();
-					type.set(i, name);
-					s.setItem(i);
+		String name = delta != null ? (String) delta[0] : null;
+		for (Object o : getExpr().getArray(e)) {
+			if (o instanceof Player) {
+				switch (mark) {
+					case 2: ((Player) o).setDisplayName(name != null ? name + ChatColor.RESET : ((Player) o).getName());
+					case 3: ((Player) o).setPlayerListName(name); // If name is null, then the player's name will be used.
 				}
-			} else {
-				Object i = getExpr().getSingle(e);
-				if (i instanceof ItemType) {
-					type.set(i, name);
-					getExpr().change(e, new ItemType[] {(ItemType) i}, ChangeMode.SET);
+			} else if (o instanceof Entity) {
+				((Entity) o).setCustomName(name);
+				if (mark == 2 || mode == ChangeMode.RESET) // Using "display name"
+					((Entity) o).setCustomNameVisible(name != null);
+				if (o instanceof LivingEntity)
+					((LivingEntity) o).setRemoveWhenFarAway(name == null);
+			} else if (o instanceof ItemType) {
+				ItemType i = (ItemType) o;
+				ItemMeta m = i.getItemMeta();
+				m.setDisplayName(name);
+				i.setItemMeta(m);
+			} else if (o instanceof Inventory ) {
+				Inventory inv = (Inventory) o;
+
+				if (inv.getViewers().isEmpty())
 					return;
-				}
+				// Create a clone to avoid a ConcurrentModificationException
+				List<HumanEntity> viewers = new ArrayList<>(inv.getViewers());

```

Comment:
Don't create a clone, use its iterator instead. Or are you even modifying the list to generate a CME? I don't see any code that could do it.

Reply:
A CME can be encountered when opening the inventory later on, at least that’s what I encountered.

Reply:
Ah, yeah, this is alright then.